### PR TITLE
feat: reduce binary size of kmeshctl

### LINKED ISSUE #946: The binary size of `kmeshctl` seems too large
**Task description**:

The binary size of `kmeshctl` as follows:

```bash
root@dev:~# ls -lh /usr/local/bin/kmeshctl 
-rwxr-xr-x 1 root root 110M Oct 11 08:11 /usr/local/bin/kmeshctl
```

In contrast, the binary size of `istioctl` is:

```bash
root@dev:~# ls -lh /usr/local/bin/istioctl 
-rwxr-xr-x 1 root root 87M Sep 24 11:54 /usr/local/bin/istioctl
```

size of `kubectl` is:

```bash
root@dev:~# ls -lh /usr/local/bin/kubectl 
-rwxr-xr-x 1 root root 46M Sep 24 07:05 /usr/local/bin/kubectl
```

size of `kmesh-daemon`:

```bash
[root@kmesh-p6cb9 kmesh]# ls -lh /usr/bin/kmesh-daemon 
-r-xr--r-- 1 root root 120M Oct  8 02:59 /usr/bin/kmesh-daemon
```

**Solution**:

Maybe optimize go compilation options or remove unnecessary dependencies?

**Who can join or take the task**:

The good first issue is intended for `first-time contributors` to get started on his/her contributor journey.

After a contributor has successfully completed 1-2 good first issue's,
they should be ready to move on to `help wanted` items, saving the remaining `good first issue` for other new contributors.

**How to join or take the task**:

Just reply on the issue with the message `/assign` in a separate line.

Then, the issue will be assigned to you.

**How to ask for help**:

If you need help or have questions, please feel free to ask on this issue.
The issue author or other members of the community will guide you through the contribution process.

### PR DESCRIPTION
**Pull request by am6737**
**What type of PR is this?**
/kind enhancement

**What this PR does / why we need it**:
This PR reduces the binary size of `kmeshctl` by replacing the dependency on `istio.io/istio/pkg/kube` and applying build flags to strip debugging symbols. 

**Which issue(s) this PR fixes**:
Closes #946

**Special notes for your reviewer**:
I noticed that `test/e2e/main_test.go` uses `istio.io/istio/pkg/kube`. Do I need to do anything about it?

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by am6737** [comment]
Hey @hzxuzhonghu 
Do you have time to take a look? I think I wrote it a bit poorly.

File: pkg/kube/client.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
+/*
```
Comment (YaoZengzeng) [inline pkg/kube/client.go:1]: All the contents of `pkg/kube/*` extracted from `istio/pkg/kube/*`? If so, I wonder if it is worthwhile to maintain so much code by ourselves just to reduce the binary size.

File: ctl/waypoint/waypoint.go [hunk 2]
```diff
@@ -37,7 +37,8 @@ import (
 	"istio.io/istio/pkg/config/constants"
 	"istio.io/istio/pkg/config/protocol"
 	"istio.io/istio/pkg/config/schema/gvk"
-	"istio.io/istio/pkg/kube"
+	"kmesh.net/kmesh/pkg/kube"
```
Comment (LiZhenCheng9527) [inline ctl/waypoint/waypoint.go:40]: this is a golint error

**Comment by YaoZengzeng** [comment]
Anyway, what is the binary size after this PR's optimization?

**Review by LiZhenCheng9527 — COMMENTED** [review]
Are you trimming down to the parts of istio kube that kmesh doesn't need? 
What is the size of the compiled `kmeshctl` compared to the original.

**Comment by am6737** [comment]
Hey @YaoZengzeng @LiZhenCheng9527 

Apologies for not clarifying this in the PR description. As mentioned in the initial submission, this change reduces the kmeshctl size from 110MB to 55MB, achieving a 50% reduction. When compiled with debug enabled, the size is around 82MB.

File: pkg/kube/client.go (hunk 1 above)
Comment (am6737) [reply pkg/kube/client.go:1]: > All the contents of `pkg/kube/*` extracted from `istio/pkg/kube/*`? If so, I wonder if it is worthwhile to maintain so much code by ourselves just to reduce the binary size.

No, not all of it, just part of the code, though it does include quite a bit. The main goal was to provide only the methods needed by ctl.

File: ctl/waypoint/waypoint.go (hunk 2 above)
Comment (am6737) [reply ctl/waypoint/waypoint.go:40]: > this is a golint error

Sorry, I can't see where the issue is. Could you help point it out specifically?

Comment (LiZhenCheng9527) [reply ctl/waypoint/waypoint.go:40]: In go, the order of packages introduced is standard pkg, tripartite pkg and packages within the organisation, with each partition split by blank lines.

You have now use `kmesh.net/kmesh/pkg/kube`, referencing a pkg within the organisation. It should be placed in the third partition

File: Makefile [hunk 3]
```diff
@@ -111,7 +123,7 @@ OUT ?= kmeshctl
 kmeshctl:
 	$(call printlog, BUILD, $(APPS4))
 	$(QUIET) (export PKG_CONFIG_PATH=$(PKG_CONFIG_PATH):$(ROOT_DIR)mk; \
-		$(GO) build -o $(OUT) $(GOFLAGS) ./ctl/main.go)
+		CGO_ENABLED=0 $(GO) build -gcflags $(GOGCFLAGS) -ldflags $(GOLDFLAGS) -o $(OUT) $(GOFLAGS) ./ctl/main.go)
```
Comment (hzxuzhonghu) [inline Makefile:126]: Keep consistent with L118

File: pkg/kube/client.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	utilruntime.Must(clientnetworkingalpha.AddToScheme(scheme))
+	utilruntime.Must(clientnetworkingbeta.AddToScheme(scheme))
+	utilruntime.Must(clientsecurity.AddToScheme(scheme))
+	utilruntime.Must(clienttelemetry.AddToScheme(scheme))
+	utilruntime.Must(clientextensions.AddToScheme(scheme))
+	utilruntime.Must(gatewayapi.Install(scheme))
+	utilruntime.Must(gatewayapibeta.Install(scheme))
+	utilruntime.Must(gatewayapiv1.Install(scheme))
+	utilruntime.Must(apiextensionsv1.AddToScheme(scheme))
+	return scheme
```
Comment (hzxuzhonghu) [inline pkg/kube/client.go:71]: Why do we need other istio apis

**Review by hzxuzhonghu — COMMENTED** [review]
One more confusing point is why do we need to intree all the client and portforwarder implementation, cannot we use raw kubectl pkg?

File: pkg/kube/client.go (hunk 4 above)
Comment (am6737) [reply pkg/kube/client.go:71]: > Why do we need other istio apis

Sorry, I missed that. Thanks for pointing it out!

**Comment by hzxuzhonghu** [comment]
>My intention was to avoid modifying the existing code logic as much as possible, so I only defined and implemented the necessary interface methods. Using the kubectl portforwarder would require more changes to the existing logic.
>
>I’d appreciate your guidance on how you think this could be improved. Thank you!

My only concern is this still have some dependency on istio. If some files are copied from istio, we must add comment where it is. And keep the original copyright.

**Comment by am6737** [comment]
> > My intention was to avoid modifying the existing code logic as much as possible, so I only defined and implemented the necessary interface methods. Using the kubectl portforwarder would require more changes to the existing logic.
> > I’d appreciate your guidance on how you think this could be improved. Thank you!
> 
> My only concern is this still have some dependency on istio. If some files are copied from istio, we must add comment where it is. And keep the original copyright.

Sorry, I didn’t think of that aspect. I realize that most of the content in client_config.go, client_factory.go, and util.go was copied from Istio. I can update the code to remove references to these three files.

**Comment by am6737** [comment]
Hey @hzxuzhonghu 

How does it look now? The code related to Istio has been removed.

**Review by hzxuzhonghu — DISMISSED** [review]

**Comment by am6737** [comment]
> Special notes for your reviewer:
I noticed that test/e2e/main_test.go uses istio.io/istio/pkg/kube. Do I need to do anything about it?

Hey @YaoZengzeng @hzxuzhonghu @LiZhenCheng9527 

Do I need to pay attention here

**Comment by hzxuzhonghu** [comment]
@am6737 That could be done later, and do not influence kmeshctl binary size

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

**Pull request merged**